packages/ui/src/theme/createTheme/resolveObject.ts

```diff
@@ -0,0 +1,148 @@
+// Internal Style Dictionary methods
+// copied from amzn/style-dictionary with the owner's permission
+import {
+  cloneDeep,
+  getName,
+  getPathFromName,
+  has,
+  resolveReference,
+  usesReference,
+} from '../../utils';
+import { REFERENCE_REGEX } from './constants';
+
+const DEFAULTS = {
+  ignoreKeys: ['original'],
+};
+
+export function resolveObject<T>(object: Record<string, any>): T {
+  const foundCirc: Record<string, boolean> = {};
+  const clone = cloneDeep(object); // This object will be edited
+  const currentContext: string[] = []; // To maintain the context to be able to test for circular definitions
+  if (typeof object === 'object') {
+    return traverseObject({
+      slice: clone,
+      fullObj: clone,
+      currentContext,
+      foundCirc,
+    }) as T;
+  } else {
+    throw new Error('Please pass an object in');
+  }
+}
+
+export function traverseObject<T>({
+  slice,
+  fullObj,
+  currentContext,
+  foundCirc,
+}): T {
+  for (let key in slice) {
+    if (!has(slice, key)) {
+      continue;
+    }
+
+    const prop = slice[key];
+    // We want to check for ignoredKeys, this is to
+    // skip over attributes that should not be
+    // mutated, like a copy of the original property.
+    if (DEFAULTS.ignoreKeys && DEFAULTS.ignoreKeys.indexOf(key) !== -1) {
+      continue;
+    }
+
+    currentContext.push(key);
+    if (typeof prop === 'object') {
+      traverseObject({ currentContext, slice: prop, fullObj, foundCirc });
+    } else {
+      if (typeof prop === 'string' && prop.indexOf('{') > -1) {
+        slice[key] = compileValue({
+          value: prop,
+          stack: [getName(currentContext)],
+          foundCirc,
+          fullObj,
+        });
+      }
+    }
+    currentContext.pop();
+  }
+
+  return fullObj as T;
+}
+
+export function compileValue({ value, stack, foundCirc, fullObj }) {
+  let toRet = value,
+    ref;
```

Comment:
For my own understanding what is this doing?

Reply:
This initializes the return value, which is the original value passed in with variables then substituted for values. and declares a variable `ref` to keep track of what the variables are referencing. Added comments at the top of the functions to explain what they do

Reply:
More curious abt the initialization syntax:
```
// makes sense, `toRet` is initialized to `value`
let toRet = value,
  //  what is happening here? Assuming `ref` initializes to `undefined`?
  ref;
```

Reply:
Yeah `ref` just starts as undefined. I think this is just shorthand to avoid doing 
```
let toRet = value
let ref;
```